In [ ]:
##############################################################################
# AI & OPTIMIZATION - Lab 04 - 2025/2026
##############################################################################

In [1]:
# System imports
import matplotlib.pyplot as plt
import numpy as np

# Extra library imports
import pandas as pd
from scipy.optimize import linprog

In [2]:
# User inputs
datafile = 'data.csv'
servfile = 'serving.csv'
bfood = ['Medium_banana', 'Instant_oatmeal']
max_val = 5
nb_points = 1000

The Dietary Approaches to Stop Hypertension (DASH) eating plan was developped by the American National Institutes of Health (NIH) to reduce blood tension. This plan favors the daily consumption of whole grains, vegetables, fruits, and low-fat protein food such as meat, poultry, and fish. More precisely, it recommands daily amount of portions for these food types to control sodium intake according to calorie needs. Examples of menus can be found online (https://www.nhlbi.nih.gov/education/dash-eating-plan).

In this lab, we will consider DASH eating planning as an optimization problem, minmizing the price of the daily selected menu.

# I. Optimization of a (quite incomplete) DASH breakfast

Suppose we want to plan a breakfast following the DASH recommandations. Let's first have a look at the food in our kitchen.

<font color='blue'> Question 1: open the 'data.csv' using pandas with the appropriate input parameters. Assign and display a variable "nb_food" with the number of different food types. Display the available information for the first three food types. </font>

In [4]:
# Load the csv file
df = pd.read_csv(datafile, sep=";")

# Display the features
nb_food = df.shape[0]
print(f'{nb_food} food products available in the kitchen.')

# Display the first three rows
df.head(3)

31 food products available in the kitchen.


,Food,Sodium_per_portion,Price_per_portion,Calories_per_portion,Group,Meal,Group_serving_per_portion
0,Almonds,1,2.15,576.00,Nuts,Lunch,0.25
1,Brocoli,11,0.17,11.33,Vegetables,Lunch,1.00
2,Brown_rice,5,0.69,359.00,Grains,Lunch,0.50


Thus, we see that one portion of almonds contains 1 mg of sodium, costs 2.15 euros and is associated with 576 kcal of calories. 

First, we want to plan a breakfast as cheap as possible. To simplify the problem, we will only choose portions of medium banana and instant oatmeal. We suppose that the number of portions to select could be any real number (for example $\frac{1}{2}$ or $\frac{2}{3}$), and that we should take at least one portion of medium banana and one portion of instant oatmeal. The breakfast should also contain between 300 and 500 kcal of calories.
We can formulate our optimal breakfast planning as a linear programming problem under the following form:

\begin{equation}
\displaystyle \min_{\boldsymbol x = (x_1, x_2)^T} \displaystyle x_1 w_1 + x_2w_2 \\ 
s.t. \begin{cases}
    c_1(x_1, x_2) \leq a_1&\\
    ... \\
     c_n(x_1, x_2) \leq a_n
  \end{cases}   
\end{equation}

with $x_1$ the number of portions of medium banana, $x_2$ the number of portions of instant oatmeal, and $n$ the number of constraints.

<font color='blue'> Question 2: what do $w_1$ and $w_2$ represent? How many constraints $c_i$ do we have? Write each constraint $c_i$ in fonction of $x_1$ and $x_2$, and write the value associated with each $a_i$.

<font color='green'> 
To complete

Since the formulated problem is in dimension 2, we can now visualize the feasible set and try to solve it visually.

<font color='blue'> Question 3: complete the following code to draw each line of equation $c_i(x_1, x_2) - a_i = 0$ and highlight the feasible set. How many basic feasible points do you see? What simple assumption can you visually make for the optimal values of $x_1$ and/or $x_2$?

In [ ]:
# Restrict the dataframe to the required food for breaksfast
df_b = 

# Get the two first food values
f1, f2 = #... to complete

# Get the two first calory values
cal_1, cal_2 = #... to complete

# Get the two first sodium values
s1, s2 = #... to complete

# Get the two first price values
p1, p2 = #... to complete

# Get the w_i values 
w1, w2 = #... to complete

# Create a vector with possible w_i values
wi_vals = np.linspace(0, max_val, nb_points)

# Create the vectors embedding possible combinations
# of w1 and w2 values
w1_vect, w2_vect = np.meshgrid(wi_vals, wi_vals)
w1_vect = w1_vect.flatten()
w2_vect = w2_vect.flatten()

# Initialize the feasible set using the created variables 
# with the first constraint:
# c1(x1, x2) - a1 <= 0
feasible_set = #... to complete

# Add the second constraint (if existing) using the created variables:
# c2(x1, x2) - a2 <= 0
feasible_set &= #... to complete if existing or to remove

# Add all the other constraints one by one (if existing)
# using the created variables:
feasible_set &= #... to complete if existing or to remove
feasible_set &= #... to complete if existing or to remove
feasible_set &= #... to complete if existing or to remove
feasible_set &= #... to complete if existing or to remove

# Initialize the figure
fig, ax = plt.subplots()

# Create the lines associated with the n constraints
# ax.plot(wi_vals, to complete, 'k')
# Add as many plot code lines as needed

# Visualize the feasible set
ax.scatter(w1_vect[feasible_set],
           w2_vect[feasible_set],
           c='y');   

# Set other figure's parameters
ax.set_xlim([0, max_val]);
ax.set_ylim([0, max_val]);
ax.set_xlabel(f1);
ax.set_ylabel(f2);     

<font color='green'> 
To complete

<font color='blue'> Question 4: automatically compute the basic feasible points, without using any additional library or complicated calculation. For each basic feasible point, display the $x_1$, $x_2$ and loss values. Create a figure to display each line of equation $c_i(x_1, x_2) - a_i = 0$, the feasible set, the basic feasible points and the optimal solution.

In [ ]:
# Initialize the figure plotting the constraints
fig, ax = plt.subplots()
# ax.plot(wi_vals, to complete, 'k')
# Add as many plot code lines as needed

# Compute the basic feasible points and the optimal solution
#... to complete

# Display the feasible set
ax.scatter(w1_vect[feasible_set],
           w2_vect[feasible_set],
           c='y');   

# Display the optimal solution
ax.plot(best_x[0],
        best_x[1],
        'or');  

# Set other figure's parameters
ax.set_xlim([0, max_val]);
ax.set_ylim([0, max_val]);
ax.set_xlabel(f1);
ax.set_ylabel(f2);     

<font color='blue'> Question 5: what initial point $\boldsymbol x^0$ can we choose to solve this problem using the simplex method? Give all the possible sequences of $\boldsymbol x^k$ points over the $k$ iterations according to the selected $\boldsymbol x^0$.

<font color='green'> 
To complete

<font color='blue'> Question 6: display the optimal portion of medium banana and instant oatmeal to take. Display the price, calory and sodium values associated with this optimal breakfast. Comment.

In [ ]:
# Display the optimal point
#... to complete

# Display the associated price
#... to complete

# Display the associated calories
#... to complete

# Display the associated sodium uptake 
#... to complete

<font color='green'> 
To complete

# II. Optimization of the DASH menu of the day

We will now create the cheapest menu for today, supposing we target between 1800 and 2000 kcal of calories and between 1500 and 2300 mg of sodium. We still supose that the optimal portions to select should be real numbers.

<font color='blue'> Question 7: create the optimal menu of the day following the above recommendations using linear programming. Display the price of the optimal menu, the number of optimal portions for each selected food product, the associated amount of sodium and calories.

In [ ]:
# Set the coefficients of the linear objective function to be minimized
c = #... to complete

# Get the calory and sodium values
cals = #... to complete
sods = #... to complete

# Get the food products
foods =  #... to complete

# Set the right part of the inequality constraints
b_ub = #... to complete

# Set the inequality constraint matrix
A_ub = #... to complete

# Perform the optimization
res = linprog(c, A_ub=A_ub, b_ub=b_ub)

# Display the optimization result
print(res.message)
if res.success == True:
    # Display the price of the optimal menu
    #... to complete

    # Display the number of optimal portions for each selected food product
    #... to complete
        
    # Display the associated amount of sodium  
    #... to complete
    
    # Display the associated amount of calories  
    #... to complete

You should see that the constraints are well respected, but that the created menu is not balanced at all! In fact, the DASH eating plan also sets a minimal and maximal daily serving for each food group. The equivalence between portion and serving is available in the column "Group_serving_per_portion" of the "data.csv" file. For example, a portion of romane lettuce is equivalent to 0.25 vegetable according to the DASH plan. Since the serving of nuts is made weekly, we do not consider this food group to optimize the menu of the day.

<font color='blue'> Question 8: load the serving.csv file. Remove the food products associated with the nuts' group in the dataframe created from the "data.csv" file. Display the remaining number of food products. Create the optimal menu of the day following the complete DASH recommendations using linear programming. Display the price of the optimal menu, the number of optimal portions for each selected food product, the associated amount of sodium and calories, and the number of servings for each food group.

In [ ]:
# Load the serving datafile
dfs =  #... to complete

# Remove the nuts from the food dataframe
dfs = #... to complete

# Display the remaining number of food products
#... to complete

# Recreate the variables impacted by the nuts' exclusion
#... to complete

# Get the serving and group information
servs = #... to complete
groups = #... to complete

# Initialize the inequality constraint matrix
A_ub = #... to complete

# Set the right part of the inequality constraints
b_ub = #... to complete

# Perform the optimization
res = linprog(c, A_ub=A_ub, b_ub=b_ub)

# Display the optimization result
print(res.message)
if res.success == True:
    # Display the price of the optimal menu
    #... to complete

    # Display the number of optimal portions for each selected food product
    #... to complete
        
    # Display the associated amount of sodium  
    #... to complete
    
    # Display the associated amount of calories  
    #... to complete
    
    # Display the number of servings for each food group
    #... to complete

We have now all the food products to select for today! However, we still have to properly compose our menu, divided into breakfast, lunch and dinner.

## III. Composition of the optimized DASH menu of the day

<font color='blue'> Question 9: get the proper menu of the day. For each meal - breakfast, lunch and dinner - display the selected food products and their optimal portions.

In [ ]:
#... to complete

You should see that the proposed daily menu is unbalanced - and more especially the breakfast. We can improve it limiting the maximal number of fruits and dairy products.

<font color='blue'> Question 10: using the bounds argument, improve the selected menu limiting the maximal number of portions according to the food product: 
 - a portion of a fruit or dairy product should be between 0 and 1,
 - a portion of a food product from any different category should be between 0 and 2.   
    
For each meal - breakfast, lunch and dinner - display the selected food products and their optimal portions.

In [ ]:
# Update the bounds argument
bounds = #... to complete

# Optimize the daily menu 
res = linprog(c, A_ub=A_ub, b_ub=b_ub, bounds=bounds)

# Display the selected meals 
#... to complete

We improved our breakfast... but we can still do better!
To have a more balanced menu, we can for example ensure that the breakfast is fairly divided into fruits, dairy products and grains.

<font color='blue'> Question 11: add equality constraints to ensure that the breakfast is equitably divided into fruits, dairy products and grains. For each meal - breakfast, lunch and dinner - display the selected food products and their optimal portions.
Display the price of the optimal menu and the associated amount of sodium and calories. Are the associated constraints respected? Comment.

In [ ]:
# Update the equality constraints
A_eq = #... to complete
b_eq = #... to complete

# Optimize the daily menu      
res = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds)

# Display the selected meals 
#... to complete

<font color='green'> 
To complete

To conclude, we will focus on the selected breakfast, and assess if it respects the constraints proposed in part I.

<font color='blue'> Question 12: display the price of the selected breakfast and the associated amount of calories. Are the constraints proposed in part I respected? Comment.

In [ ]:
# To complete

<font color='green'> 
To complete